drivers: digital-io: max149x6 : Add max14916 driver support

RaduSabau1 opened the pull request:
## Pull Request Description

MAX14906 and MAX14916 are similar chips, so the crc, read and write functions are the same, therefore those exact API's
and other common macro's are moved to a base driver which is to be used in the max14906 driver as well as the max14916
driver.
Add max14916 driver inital header and source files, which as mentioned before includes the max149x6-base driver.
Add max14916 iio driver initial header and source files.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

amiclaus reviewed (commented):
is this PR subject for review yet?

RaduSabau1 commented:
> is this PR subject for review yet?

Not tested with hardware yet.

RaduSabau1 commented:
V2

- Activating all chanels at initialization now
- Errors return error codes now

RaduSabau1 commented:
V3

- Removed climit_get function since it was a simple reg_read.
- Added sled_set function to set the state of a specific status led.

amiclaus reviewed (commented):
this approach is a bit unusual, looking at other drivers that have `-base` source/header files on no-os:
https://github.com/analogdevicesinc/no-OS/blob/master/drivers/adc-dac/ad5592r/ad5592r-base.c
or linux:
https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/accel/adxl345_core.c
https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/ad7091r-base.c

are used to store common data regarding the part specification/functionality, while the specific part source files handle usually the type of communication/extra channels/extra registers.

In your case it is exactly vice-versa, all the registers and functionality are different, while the only common thing is the communication (hopefully I am not mistaken). Is it worth adding extra source files only for some basic read/write functions?

buha commented on drivers/digital-io/max149x6/max149x6-base.h:
```diff
@@ -0,0 +1,87 @@
+/**
+ * @brief Initialization parameter for the MAX149X6 device.
+ */
+struct max14906_init_param {
```
why not max149x6_init_param ?

RaduSabau1 replied:
will change to max149x6

buha commented on drivers/digital-io/max149x6/max149x6-base.h:
```diff
@@ -0,0 +1,87 @@
+/**
+ * @brief Device descriptor for MAX149X6.
+ */
+struct max14906_desc {
```
same here, why not max149x6_desc ?

CiprianRegus replied:
This was my idea to not break existing projects. But now, other projects require makefile changes... so I don't know if it's worth keeping this naming.

RaduSabau1 replied:
will change to max149x6_desc after all.

buha commented:
the last commit is mixed up and made of renames and edits

for clarity i would do a commit with renaming only right after 2nd commit

and then the 4th and 5th would be adding max14916 and iio_max14916 respectively

CiprianRegus reviewed (changes requested):
This will break the swiot1l project. Rebase on master and update the required makefiles.

CiprianRegus commented on drivers/digital-io/max14916/max14916.c:
```diff
@@ -0,0 +1,268 @@
+	int ret;
+
+	/* FLEDs need to be controlled internally which requires FLEDSet bit
+	   in the config1 rgister to be set to 0. */
```
register

RaduSabau1 replied:
will fix it.

CiprianRegus commented on drivers/digital-io/max149x6/iio_max14916.c:
```diff
@@ -0,0 +1,322 @@
+		.name = "scale",
+		.show = max14916_iio_read_scale,
+	},
+	{
+		.name = "config2",
+		.show = max14916_iio_read_config2,
+		.store = max14916_iio_write_config2,
+	},
```
this should not be an attribute

RaduSabau1 replied:
will remove.

RaduSabau1 commented on drivers/digital-io/max14916/max14916.c:
```diff
@@ -0,0 +1,268 @@
+	if (ch >= MAX14916_CHANNELS)
+		return -EINVAL;
+
+	if (val < 0 || val > 1)
```
this is my bad, will fix it.

RaduSabau1 commented on drivers/digital-io/max14916/max14916.c:
```diff
@@ -0,0 +1,268 @@
+ * @param wd - watchdog timer to be selected.
+ * @return 0 in case of succes, negative error code otherwise.
+*/
+int max14916_set_config2(struct max14906_desc *desc,
```
will split it into 3 separate functions

RaduSabau1 commented:
V4

- Modified the commit tree
- Renamed descriptor and init_param in the base driver files and all other files affected by the change.
- Addressed rest of the comments.

RaduSabau1 commented:
V5

- Fixed  typos, rebased master and updated the staging/max14916_driver branch.
- Made changes in the makefile for the swiot1l project.

buha commented on projects/swiot1l/src.mk:
```diff
@@ -58,17 +58,17 @@ SRCS += $(DRIVERS)/api/no_os_gpio.c \
 
-INCS += $(DRIVERS)/digital-io/max14906/max14906.h
-SRCS += $(DRIVERS)/digital-io/max14906/max14906.c
+INCS += $(DRIVERS)/digital-io/max149x6/max14906.h
```
this is not enough, some .c files from this project need to be modified as well to use the max149x6 desc and init_param

buha replied:
and i would merge this change with the commit where you added "base" functionality

RaduSabau1 replied:
On it.

CiprianRegus replied:
Still not resolved

buha commented on drivers/digital-io/max149x6/iio_max14906.c:
```diff
@@ -255,11 +255,11 @@ static int max14906_iio_read_do_mode(void *dev, char *buf, uint32_t len,
 				     const struct iio_ch_info *channel, intptr_t priv)
 {
 	struct max14906_iio_desc *iio_desc = dev;
-	struct max14906_desc *desc = iio_desc->max14906_desc;
```
that commit title is too long, you don't need to put the whole file hierarchy in the commit title, just put the chip name and it's clear enough

if you want to get the files changed by a commit, it's easily doable with `git show [hash] --name-only`

RaduSabau1 replied:
Making the commit title shorter.

RaduSabau1 commented:
V6

- swiot1l project src.mk has updated include path
- shortened third's commit title.

CiprianRegus commented on drivers/digital-io/max149x6/iio_max14916.c:
```diff
@@ -0,0 +1,256 @@
+/***************************************************************************//**
```
Please test this

RaduSabau1 replied:
i will look again at the code for the iio driver, i noticed other bugs myself.

RaduSabau1 replied:
I tested it, and a bool array ch_enable created by the user and transmitted through the iio_desc_init_param tells the iio_init function where the iio_setup_channels is called which channels will be active, and works when using iio oscilloscope to see the active channels.

RaduSabau1 commented on drivers/digital-io/max149x6/iio_max14916.c:
```diff
@@ -0,0 +1,256 @@
+				   const struct iio_ch_info *channel,
+				   intptr_t priv)
+{
+	int32_t val = 0;
```
Fixing it.

RaduSabau1 commented:
V7

- Removed unnecessary functions or variables.
- Added iio_setup_channels function and bool array on the iio descriptor and init param.
- Modified swiot1l project to address build errors.

RaduSabau1 commented:
V8

- Modified commits.

RaduSabau1 commented:
V9

- Made changes to iio driver.

RaduSabau1 commented:
V10

- modified iio_setup_channels in order to fix a memory leak.

CiprianRegus commented on drivers/digital-io/max149x6/iio_max14916.h:
```diff
@@ -0,0 +1,71 @@
+struct max14916_iio_desc {
+	struct max149x6_desc *max14916_desc;
+	struct iio_device *iio_dev;
+	uint32_t active_channels;
+	uint32_t no_of_active_channels;
```
This is not resolved, these 2 fields are unused.

RaduSabau1 replied:
will remove them.

CiprianRegus commented on drivers/digital-io/max149x6/iio_max14916.c:
```diff
@@ -0,0 +1,296 @@
+			max14916_iio_channels[ch_offset].attributes = max14916_attrs;
+			ch_offset++;
+		}
+		max14916_iio_channels[i].ch_out = 1;
```
you're accessing potentially unallocated memory

RaduSabau1 replied:
moving this line in the if condition above.

CiprianRegus commented on drivers/digital-io/max149x6/iio_max14916.c:
```diff
@@ -0,0 +1,296 @@
+
+	descriptor->iio_dev = &max14916_iio_dev;
+
+	memcpy(descriptor->ch_enable, init_param->ch_enable,
```
You don't need to store `ch_enable` in the descriptor, as it's only used during init

RaduSabau1 replied:
ch_enable will be stored only in the init_param since it is not needed only during init.

RaduSabau1 commented:
V11

- Removed ch_enable array from iio descriptor
- Fixed usage of potentially unallocated memory,

RaduSabau1 commented:
V12

- Removed unused variables.

RaduSabau1 commented:
V13

- Added GPIOs for READY, FAULT and SYNCH.

CiprianRegus commented:
The last change requires a separate commit.

RaduSabau1 commented:
V14

- Organized commits properly.

CiprianRegus commented on drivers/digital-io/max149x6/max14906.c:
```diff
@@ -199,6 +199,40 @@ int max14906_init(struct max149x6_desc **desc,
+		goto spi_err;
+
+	if (descriptor->synch_gpio) {
+		ret = no_os_gpio_set_value(descriptor->synch_gpio,
```
You have to set this gpio as output first

RaduSabau1 replied:
Fixing it.

RaduSabau1 commented:
V15

- Fixed Synch GPIO initialization.

RaduSabau1 commented:
V16

- Added GPIO changes to MAX14916 as well.
- Added README.rst

RaduSabau1 commented:
V17

- Fix README.rst code block.

buha commented on drivers/digital-io/max149x6/max14906.c:
```diff
@@ -199,6 +199,40 @@ int max14906_init(struct max149x6_desc **desc,
+	}
+
+	ret = no_os_gpio_get_optional(&descriptor->ready_gpio,
+				      param->synch_gpio_param);
```
use consistent naming

buha commented on drivers/digital-io/max149x6/max14916.c:
```diff
@@ -235,6 +235,40 @@ int max14916_init(struct max149x6_desc **desc,
+	}
+
+	ret = no_os_gpio_get_optional(&descriptor->ready_gpio,
+				      param->synch_gpio_param);
```
same here

RaduSabau1 commented:
V18

- Updated README
- Fix gpio bad naming issue.

buha commented on drivers/digital-io/max149x6/README.rst:
```diff
@@ -0,0 +1,322 @@
+
+MAX14916 has only 3 attributes :
+
+raw - indicates the state of a channel(0 or 1).
```
these 3 don't render as you would like, they should probably be listed with bullets

RaduSabau1 commented:
V19

- Updated README.

CiprianRegus reviewed (approved):
Tested the max14906 with swiot1l. Works as expected.

buha reviewed (approved):
ok, i think we can merge this
